### PR TITLE
JAVA-1970: Expose a counter for in-flight speculative executions

### PR DESCRIPTION
**Pull request by adutra**

### CODE REVIEW

File: core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandlerBase.java [hunk 1]
```diff
@@ -755,13 +764,24 @@ public void cancel() {
...
       context
           .getRequestTracker()
           .onNodeError(statement, error, latencyNanos, executionProfile, node);
     }
 
+    private void maybeUpdateSpeculativeExecutionsInFlightCounter(int amount) {
+      if (this.execution > 0) {
+        ((DefaultNode) node)
+            .getMetricUpdater()
+            .incrementCounter(
```
Comment (adutra) [inline core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandlerBase.java:778]: Out of curiosity: what is the reason why we pass the profile name to all methods related to metrics? It's not used in the default implementation.

File: core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandlerBase.java [hunk 2]
```diff
@@ -755,13 +764,24 @@ public void cancel() {
...
-    private void trackNodeError(Node node, Throwable error) {
+    private void trackNodeError(Throwable error) {
       long latencyNanos = System.nanoTime() - this.start;
       context
           .getRequestTracker()
           .onNodeError(statement, error, latencyNanos, executionProfile, node);
     }
 
+    private void maybeUpdateSpeculativeExecutionsInFlightCounter(int amount) {
+      if (this.execution > 0) {
```
Comment (tomekl007) [inline core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandlerBase.java:775]: why we need that conditional logic here? How can be a situation when we don't have execution > 0 but the execution finished executing: `maybeUpdateSpeculativeExecutionsInFlightCounter`?

File: core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandlerBase.java (hunk 1 above)
Comment (olim7t) [reply core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandlerBase.java:778]: I thought about splitting the metrics by execution profile at some point. I decided not to do it, but didn't want to completely close the door to the idea.

**Review by olim7t — COMMENTED** [review]
I'm not convinced of the usefulness of this new metric.

`speculative-executions.triggered` (the old one) counts the speculative executions triggered by a slow response from this node. As an operator, if I see it going up, it's a sign that this particular node is too slow for the configured threshold, which would prompt me to further investigate performance and/or adjust the threshold.

`speculative-executions.in-flight`, on the other hand, tells me how many speculative executions _triggered by other nodes_ are running on this node. How is that useful? If I see a high number, what does that tell me about this node? Nothing really, just that the other nodes are slower, which I can already see with the other metric.

File: core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandlerBase.java (hunk 2 above)
Comment (adutra) [reply core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandlerBase.java:775]: If `execution == 0` then this is the first (i.e., non-speculative execution of the request); for `execution > 0`, we are sure that this is a speculative execution.

**Comment by adutra** [comment]
Indeed there is no internal need for this metric anymore atm. Closing.

**Pull request closed without merging**